### DISCUSSION TITLE
Proper way to restore the expand/collapse state of virtual viewers

### DISCUSSION THREAD

**Discussion by ptziegler** [eclipse-platform/eclipse.platform.ui#2491] [General]
This is an issue I've been scratching my head at for the past few days... I have a virtual tree viewer where I fetch the currently expanded elements via `getExpandedElements()`, update the input and then restore it via `setExpandedElements()`. 

The problem: This doesn't work, as updating the input collapses the tree and therefore all tree items are virtual and without data object. Restoring the expanded state ist herefore ignored (line 2175).

https://github.com/eclipse-platform/eclipse.platform.ui/blob/9474c952a23a2c898ba25fd8a09d498434bca0af/bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java#L2170-L2190

For my next attempt, I tried to use `getTreePathFromItem()`, but this has a similar issue except that it simply throws an exception because `segment == null` (line 3189).

https://github.com/eclipse-platform/eclipse.platform.ui/blob/9474c952a23a2c898ba25fd8a09d498434bca0af/bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java#L3185-L3194

After overwriting this method and simply returning null instead of throwing an exception, the next error was thrown because the method `getExpandedTreePaths()`  skips such null elements but doesn't consider this when calculating the size of the returned array (line 3181).

https://github.com/eclipse-platform/eclipse.platform.ui/blob/9474c952a23a2c898ba25fd8a09d498434bca0af/bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/AbstractTreeViewer.java#L3271-L3282

My question: Some of those checks seem intentional, so is this functionality something that's generally unsupported for virtual trees and something where I have to switch back to a "normal" tree?

**Comment by laeubi** [comment]
Virtual Trees are a bit tricky what I do in one of my applications with a virtual tree is to start with a `TreePath` containing the root object (that is collapsed), then I expand that one then use `path.createChildPath(nextSegment)` expand that and so on until I have expanded all items.

As expanding one level of the tree then loads the children, this iteratively expands everything up to the desired state.

**Comment by ptziegler** [comment]
> what I do in one of my applications with a virtual tree is to start with a TreePath containing the root object (that is collapsed), then I expand that one then use path.createChildPath(nextSegment) expand that and so on until I have expanded all items.

That sounds similar to my latest approach, except that I need to rely on `getTreePathFromItem()` because I want to restore the previous selection and therefore don't know which segments to expand.

As a side node: I've made some smaller changes to the problematic methods  and then `setExpandedTreePaths()` works as I would expect. Would this be worth being push back upstream?

**Reply by laeubi** [reply]
Any improvements are welcome in that area sadly virtual table/tree are not very often used in platform even though they are powerful and offer descent performance.

But in general restoring a selection can become quite tricky if you do not know the path because that could mean you need to possibly load the whole data-model what obviously is not what one wants. Therefore you should carefully check how you can enhance your objects with a way to identify them in such a scenario where you replace the full model, but that's of course hard to tell without explicit examples.

Other than that, you might want to check if you can potentially compute a *delta* instead and only refresh portions of your tree instead of replace the whole (expensive) refresh operation at all.

**Reply by ptziegler** [reply]
I think I really have to sit down and investigate how exactly this issue can be reproduced, but right now it seems impossible for me to create a small reproducer...

But from what I can tell, this issue is related to https://bugs.eclipse.org/bugs/show_bug.cgi?id=287765 as it only happens when `setUseHashlookup()` is enabled and only shows up when parts of the expanded tree items are outside the visible area and thus still virtual.

Weirdly enough, if I call `getExpandedElements()` before `getExpandedTreePaths()`, no error occurs. So this method seems to force an update for the virtual widgets... Something is definitily going wrong, I just can't pinpoint it. And until then, I won't be able to bring anything constructive to the table. ☹️

**Reply by laeubi** [reply]
`setUseHashlookup` is quite mandatory for virtual table/trees.... Maybe you can enhance some of the snippets? A while back I created
- https://github.com/eclipse-platform/eclipse.platform.ui/pull/976

but it all really depends on your datamodel and probably state of virtual items, maybe you also want more the `TreeSelection` than the expanded elements?

**Reply by ptziegler** [reply]
> Maybe you can enhance some of the snippets?

Good idea. Though I'll probably create a new snippet instead.

In any case, I finally figured out what's causing the problem... 🎉 But I suppose it's easier to understand once the example is ready.

> but it all really depends on your datamodel and probably state of virtual items, maybe you also want more the TreeSelection than the expanded elements?

While the TreeSelection also works with TreePaths, I think it serves a different purpose. What I actually want is to add/remove an element from the tree without loosing the expanded state.

Normally, I would do this by modifying the model and then simply calling `viewer.refresh()` but sadly it's not as simple in this particular case. Which leaves me with a hideous workaround similar to this:

```java
TreePath[] expandedState = viewer.getExpandedTreePaths();
viewer.setInput(newInput);
viewer.setExpandedTreePaths(expandedState);
```

Where the old and new input only differ by a single element.

**Reply by laeubi** [reply]
> Normally, I would do this by modifying the model and then simply calling viewer.refresh()

This would refresh **everything** and throw away all state and makes updates very expensive, if you remove an item from the tree/model better use `org.eclipse.jface.viewers.AbstractTreeViewer.remove(Object, Object...)`, for adding an item use `org.eclipse.jface.viewers.AbstractTreeViewer.add(Object, Object...)`.

**Reply by ptziegler** [reply]
I know... but as I said: It's not that simple.

I don't actually know which element was modified without calculating a full diff and I also don't know whether it has been added or removed (which I suppose the diff would then tell me...). And that's sadly not something that can be changed without breaking apart everything that is attached to this viewer.

**Reply by laeubi** [reply]
> I know... but as I said: It's not that simple.

As said lazy models are powerful but require some good design, and I suspect something is strange with your data-model as you obviously know _that_ something has changed so why not carry the _what changed_ together in your notification?

Anyways as mentioned this will result in devastating performance compared to a true data-model backed approach where a datamodel is the place where modifications happens and then properly updates the UI. So if we talk about "proper ways" I would not find it proper to refresh the whole viewer of a lazy model as it a bit defeats its purpose.

**Reply by ptziegler** [reply]
> As said lazy models are powerful but require some good design, and I suspect something is strange with your data-model as you obviously know that something has changed so why not carry the what changed together in your notification?

That viewer and data model was designed back in the Eclipse 2 era and for such a small scope where it didn't really matter if one refreshes the entire tree. So a proper design has sadly not been the highest priority. If it were designed today, obviously one could use more modern design patterns.

But the harsh truth is that the world runs on legacy code and at some point I have to make compromises regarding which parts can be cleaned up and which can't.

**Reply by laeubi** [reply]
> But the harsh truth is that the world runs on legacy code and at some point I have to make compromises regarding which parts can be cleaned up and which can't.

The good part is that it somehow forces you to improve the LazyTree support :-D

>  it didn't really matter if one refreshes the entire tree

Maybe then a LazyTree is not even required?!

**Reply by ptziegler** [reply]
> The good part is that it somehow forces you to improve the LazyTree support :-D

I at least hope so... While I was able to reproduce the problem reliably, I still have a hard time understanding _why_ it happens and this whole virtual/non-virtual distinction in the viewer doesn't help in making this any less confusing. :/

> Maybe then a LazyTree is not even required?!

Probably... I think performance-wise,  I think they break pretty much even. Though I still have the slim chance that at some point, I get more time to have this viewer actually utilize some of the lazy-loading mechanisms.

Regardless, I still have a more recent viewer to work on where I actually _do_ know what changes. So your suggestions are very much appreciated. :-)

**Reply by laeubi** [reply]
I think I somewhere started some work to make using lazy viewers more convenient but can't find it right now, at least one really needs to rethink some common usages, sadly JFace lacks a real model (even though everywhere "model" is mentioned) compared to e.g. Swing.

**Comment by vogella** [comment]
> Would this be worth being push back upstream?

Yes.
